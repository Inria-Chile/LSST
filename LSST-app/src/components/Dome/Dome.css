.dome-container {
    display: flex;
    flex-direction: column;
    margin: 5px;
    height: 100%;
}

.dome-container .draggable-title {
    display: block;
}

.dome-container h2 {
    margin: 0px;
    margin-bottom: 5px;
    color: #fff;
}

.dome-container h4 {
    background: #3a4d61;
    font-weight: normal;
}

.dome-container .container {
    background: #3a4d61;
    width: 100%;
    margin: 0px;
    height: 100%;
    max-width: none;
}

.dome-position-container {
    vertical-align: top;
    margin: 0px 40px;
    display: flex;
    align-items: center;
    justify-content: space-evenly;
    margin: 0px;
    padding: 20px;
    border-bottom: var(--separator-size) solid var(--theme-separator);
}

.dome-azimuth-time-series-container {
    vertical-align: top;
    margin: 0px 40px;
    display: flex;
    align-items: center;
    justify-content: space-evenly;
    margin: 10px;
    padding: 20px;
    border-top: var(--separator-size) solid var(--theme-separator);
}

.m1m3-heading-col {
    padding: 0px;
    padding-left: 5px;
}

.m2-heading-col {
    padding: 0px;
}

.mirror-row {
    margin-bottom: 50px;
}

.mirror-bottom-row {
    margin-bottom: 25px;
}

.time-series-container {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;    
}

.time-series-container h5 {
    padding: 30px 0px;
}

.dome-perspectives-container {
    border-bottom: var(--separator-size) solid var(--theme-separator);
}

.timeseries-legend {
    display: flex;
    justify-content: center;
}