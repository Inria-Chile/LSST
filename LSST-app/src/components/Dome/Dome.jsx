import React, { PureComponent } from 'react';
// import ReactDOM from 'react-dom';
import DomePosition from './DomePosition/DomePosition';
import DomePerspectives from './DomePerspectives/DomePerspectives';
import DomeAzimuthTimeSeries from './TimeSeries/DomeAzimuthTimeSeries/DomeAzimuthTimeSeries';
import DomeElevationTimeSeries from './TimeSeries/DomeElevationTimeSeries/DomeElevationTimeSeries';
import DraggableTitle from '../Utils/DraggableTitle';
import TimeSeriesLegend from './TimeSeries/TimeSeriesLegend';
import './Dome.css';
import openSocket from 'socket.io-client';
import {DEMO_MODE} from '../Utils/Utils';

class Dome extends PureComponent {

    static viewName = 'dome';

    constructor(props) {
        super(props);
        this.state = {
            shuttersAperture: 0,
            domeAzimuth: 0,
            domeTargetAzimuth: 0,
            domeOptimalAzimuth: 0,
            telescopeElevation: 0,
            telescopeTargetElevation: 0,
            telescopeOptimalElevation: 0,
            mountAzimuth: 0,
        }
        this.socket = openSocket(window.location.origin + '/domeposition');      
        // this.socket.join('test');
        // console.log('SCOKERT', this.socket.on('data', msg => console.log('data', msg)));
        // console.log('SCOKERT', this.socket.on('Rotator', msg => console.log('Rotator', msg)));
        // console.log('SCOKERT', this.socket.on('Louvers', msg => console.log('Louvers', msg)));
        console.log('DomePosition\'s socket listening', this.socket.on('DomePosition', msg => this.receiveDomePositionData(msg)));
        // console.log('SCOKERT', this.socket.on('DomeShutter', msg => console.log('DomeShutter', msg)));
    }

    receiveDomePositionData = (msg) => {
        // console.log('receiveDomePositionData', msg);
        this.setState({
            domeAzimuth: msg.DomeAzPos,
            telescopeElevation: msg.DomeElPos,
            domeTargetAzimuth: msg.DomeAzCMD,
            telescopeTargetElevation: msg.DomeElCMD,
            mountAzimuth: msg.DomeAzPos,
        });
    }

    updateShuttersAperture = (aperture) => {
        this.setState({
            shuttersAperture: aperture
        });
    }

    updateDomePos = (az, el) => {
        this.setState({
            domeAzimuth: az,
            telescopeElevation: el
        })
    }

    componentDidUpdate(prevProps, prevState) {
    }

    normalizeAngle = (angle, max) => {
        return ((angle % max) + max) % max;
    }

    componentDidMount() {
        if(DEMO_MODE){
            setInterval(() => {
                let newAzimuth = this.normalizeAngle(this.state.domeTargetAzimuth, 360);
                let domeTargetAzimuth = this.normalizeAngle(this.state.domeTargetAzimuth + ((Math.random() - 0.1) * 20), 360);
                let domeOptimalAzimuth = this.normalizeAngle(newAzimuth + ((Math.random() - 0.1) * 6), 360);
                
                let newElevation = this.normalizeAngle(this.state.telescopeTargetElevation, 90);
                let telescopeTargetElevation = this.normalizeAngle((Math.random() - 0.1) * 15, 90);
                let telescopeOptimalElevation = this.normalizeAngle(newElevation + Math.random() * 3, 90);
                let newMountAzimuth = newAzimuth + ((Math.random() - 0.1) * 30);
                this.setState({
                    domeAzimuth: newAzimuth,
                    domeTargetAzimuth: domeTargetAzimuth,
                    domeOptimalAzimuth: domeOptimalAzimuth,
                    telescopeElevation: newElevation,
                    telescopeTargetElevation: telescopeTargetElevation,
                    telescopeOptimalElevation: telescopeOptimalElevation,
                    mountAzimuth: newMountAzimuth
                })
            }, 2000)
        }
    }

    render() {
        let legendData = [
            { label: 'Current value', color: '#ddd', lineDash: 0 }, 
            { label: 'Target', color: '#5e8ba9', lineDash: 0 }, 
            { label: 'Optimal', color: '#3e6b89', lineDash: 7 }, 
        ];
        
        return (
            <div className="dome-container">
                <DraggableTitle title='Dome'/>
                <div className="container pull-left">
                    <div className="row">
                        <div className="col-12">
                            <DomePosition width={330} height={300} id="dome-top" scale={1.4} xOffset={-0.05} yOffset={0.15}
                                shuttersAperture={this.state.shuttersAperture} domeAzimuth={this.state.domeAzimuth}
                                telescopeElevation={this.state.telescopeElevation} mountAzimuth={this.state.mountAzimuth} updateDomePos={this.updateDomePos} />
                        </div>
                    </div>
                    <DomePerspectives shuttersAperture={this.state.shuttersAperture} updateShuttersAperture={this.updateShuttersAperture} />
                    <div className='time-series-container'>
                        <div>
                            <h5>Azimuth</h5>
                            <DomeAzimuthTimeSeries domeAzimuth={this.state.domeAzimuth}
                                domeTargetAzimuth={this.state.domeTargetAzimuth}
                                domeOptimalAzimuth={this.state.domeOptimalAzimuth}
                                width={600} height={350} />
                        </div>
                        <div>
                            <h5>Elevation</h5>
                            <DomeElevationTimeSeries telescopeElevation={this.state.telescopeElevation}
                                telescopeTargetElevation={this.state.telescopeTargetElevation}
                                telescopeOptimalElevation={this.state.telescopeOptimalElevation}
                                width={600} height={350} />
                        </div>
                    </div>
                        <TimeSeriesLegend legendData={legendData}/>
                </div>
            </div>
        );
    }
}

export default Dome;
