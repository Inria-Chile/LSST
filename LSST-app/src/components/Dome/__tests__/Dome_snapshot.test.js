import React from 'react';
import Dome from '../Dome';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {shallow} from "enzyme"; 
import toJson from 'enzyme-to-json';
import SocketMock from 'socket.io-mock'
Enzyme.configure({ adapter: new Adapter() });


/*---- Snapshot testing ----*/
describe('Dome test', function(){
    let socket,message,dome;
    beforeEach(()=>{
        dome = shallow(
            <Dome/>);
        let msg = {
            domeAzimuth: 0,
            telescopeElevation: 0,
            domeTargetAzimuth: 0,
            telescopeTargetElevation:0,
            mountAzimuth: 0,
            timestamp: "2018-01-04T19:44:10.611Z",
        }
        socket = new SocketMock();
        socket.on('DomePosition',msg=>{
            dome.setState({
                domeAzimuth: msg.DomeAzPos,
                telescopeElevation: msg.DomeElPos,
                domeTargetAzimuth: msg.DomeAzCMD,
                telescopeTargetElevation: msg.DomeElCMD,
                mountAzimuth: msg.DomeAzPos,
                timestamp: msg.timestamp,
            })
        });
        message={
            DomeAzPos : 0,
            DomeElPos : 0,
            DomeAzCMD : 0,
            DomeElCMD : 0,
            DomeAzPos : 0,
            timestamp : "2018-01-04T19:44:10.611Z",
        }
        socket.socketClient.emit('DomePosition',message);
    });
    it('renders correctly with shallowRender',()=>{

        expect(toJson(dome)).toMatchSnapshot(); 
    })
})