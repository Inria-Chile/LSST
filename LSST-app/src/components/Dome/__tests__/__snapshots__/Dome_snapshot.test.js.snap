// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dome test renders correctly with shallowRender 1`] = `
<div
  className="dome-container"
>
  <DraggableTitle
    title="Dome"
  />
  <div
    className="container pull-left"
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <DomePosition
          domeAzimuth={0}
          height={300}
          id="dome-top"
          mountAzimuth={0}
          scale={1.4}
          shuttersAperture={null}
          telescopeElevation={0}
          updateDomePos={[Function]}
          width={330}
          xOffset={-0.05}
          yOffset={0.15}
        />
      </div>
    </div>
    <DomePerspectives
      shuttersAperture={null}
      updateShuttersAperture={[Function]}
    />
    <div
      className="time-series-container"
    >
      <div>
        <h5>
          Azimuth
        </h5>
        <DomeAzimuthTimeSeries
          domeAzimuth={0}
          domeOptimalAzimuth={null}
          domeTargetAzimuth={0}
          height={350}
          timestamp="2018-01-04T19:44:10.611Z"
          width={600}
        />
      </div>
      <div>
        <h5>
          Elevation
        </h5>
        <DomeElevationTimeSeries
          height={350}
          telescopeElevation={0}
          telescopeOptimalElevation={null}
          telescopeTargetElevation={0}
          timestamp="2018-01-04T19:44:10.611Z"
          width={600}
        />
      </div>
    </div>
    <TimeSeriesLegend
      legendData={
        Array [
          Object {
            "color": "#ddd",
            "label": "Current value",
            "lineDash": 0,
          },
          Object {
            "color": "#5e8ba9",
            "label": "Target",
            "lineDash": 0,
          },
          Object {
            "color": "#3e6b89",
            "label": "Optimal",
            "lineDash": 7,
          },
        ]
      }
    />
  </div>
</div>
`;
